mrjob/dataproc.py

```diff
@@ -0,0 +1,873 @@
+            "isPreemptible": False,
+        }
+
+        # TODO - mtai @ davidmarin - Add support for secondary worker configs
+        # Compute ONLY
+        secondary_worker_instance_group_config = {
+            "numInstances": 0,
+            "machineTypeUri": machine_uri,
+            "isPreemptible": True,
+        }
+
+        cluster_data = {
+            'projectId': self._gcp_project,
+            'clusterName': self._gcp_cluster,
+            'config': {
+                'gceClusterConfig': {
+                    'zoneUri': zone_uri
+                },
+                'masterConfig': master_instance_group_config,
+                'workerConfig': worker_instance_group_config,
+                # 'secondaryWorkerConfig': secondary_worker_instance_group_config,
+                'initializationActions': [
+                    dict(executableFile=bootstrap_cmd) for bootstrap_cmd in bootstrap_cmds
+                ]
+            }
+        }
+
+        self.dataproc_client.clusters().create(
+            projectId=self._gcp_project,
+            region=self._gcp_region,
+            body=cluster_data).execute()
+        cluster_state = None
+
+        # TODO - mtai @ davidmarin - See https://cloud.google.com/dataproc/reference/rest/v1/projects.regions.clusters#State
+        # TODO - mtai @ davidmarin - We should probably pull these strings to constants
+
+        # TODO - mtai @ davidmarin - Can end up in infinite loop if cluster_state doesn't end up in (RUNNING,ERROR)
+        while bool(cluster_state != 'RUNNING'):
+            result_describe =  self.dataproc_client.clusters().get(projectId=self._gcp_project,
+                                               region=self._gcp_region,
+                                               clusterName=self._gcp_cluster).execute()
+    
+            cluster_state = result_describe['status']['state']
+            if cluster_state == 'ERROR':
+                raise Exception(result_describe)
+    
+            log.info('Waiting for cluster to launch... sleeping 5 second(s)')
+            time.sleep(5.0)
+
+        # TODO - mtai @ davidmarin - We should probably do something nicer here...
+        assert cluster_state == 'RUNNING'
+        return self._gcp_project
+    
+    def _delete_cluster(self):
+        return self.dataproc_client.clusters().delete(
+            projectId=self._gcp_project,
+            region=self._gcp_region,
+            clusterName=self._gcp_cluster
+        ).execute()
+
+    def _get_job(self, job_id):
+        return self.dataproc_client.jobs().get(
+            projectId=self._gcp_project,
+            region=self._gcp_region,
+            jobId=job_id
+        ).execute()
+    
+    def _cancel_job(self, job_id):
+        return self.dataproc_client.jobs().cancel(
+            projectId=self._gcp_project,
+            region=self._gcp_region,
+            jobId=job_id
+        ).execute()
+
+    def _delete_job(self, job_id):
```

Comment:
Oh, you can just delete a job. That's convenient! Might as well implement `_cleanup_job()` then.